libsolidity/analysis/ControlFlowBuilder.cpp

```diff
@@ -363,18 +364,151 @@ bool ControlFlowBuilder::visit(FunctionTypeName const& _functionTypeName)
+}
+
+void ControlFlowBuilder::operator()(yul::ForLoop const& _forLoop)
+{
+	solAssert(m_currentNode && m_inlineAssemblyAnnotation, "");
+
+	(*this)(_forLoop.pre);
+
+	auto condition = createLabelHere();
+
+	if (_forLoop.condition)
+		visit(*_forLoop.condition);
+
+	auto loopExpression = newLabel();
+	auto nodes = splitFlow<2>();
+	auto afterFor = nodes[1];
+	m_currentNode = nodes[0];
+
+	{
+		BreakContinueScope scope(*this, afterFor, loopExpression);
+		(*this)(_forLoop.body);
+	}
+
+	placeAndConnectLabel(loopExpression);
+
+	(*this)(_forLoop.post);
+
+	connect(m_currentNode, condition);
+	m_currentNode = afterFor;
+}
+
+void ControlFlowBuilder::operator()(yul::Break const&)
+{
+	solAssert(m_currentNode && m_inlineAssemblyAnnotation, "");
+	solAssert(m_breakJump, "");
+	connect(m_currentNode, m_breakJump);
+	m_currentNode = newLabel();
+}
+
+void ControlFlowBuilder::operator()(yul::Continue const&)
+{
+	solAssert(m_currentNode && m_inlineAssemblyAnnotation, "");
+	solAssert(m_continueJump, "");
+	connect(m_currentNode, m_continueJump);
+	m_currentNode = newLabel();
+}
+
+void ControlFlowBuilder::operator()(yul::Identifier const& _identifier)
+{
+	solAssert(m_currentNode && m_inlineAssemblyAnnotation, "");
+	if (m_inlineAssemblyAnnotation->externalReferences.count(&_identifier))
 	{
-		if (auto variableDeclaration = dynamic_cast<VariableDeclaration const*>(ref.second.declaration))
+		if (auto const* declaration = dynamic_cast<VariableDeclaration const*>(m_inlineAssemblyAnnotation->externalReferences.at(&_identifier).declaration))
 			m_currentNode->variableOccurrences.emplace_back(
-				*variableDeclaration,
-				VariableOccurrence::Kind::InlineAssembly,
-				&_inlineAssembly
+				*declaration,
+				VariableOccurrence::Kind::Access,
+				_identifier.location
 			);
 	}
-	return true;
+}
+
+void ControlFlowBuilder::operator()(yul::Assignment const& _assignment)
+{
+	solAssert(m_currentNode && m_inlineAssemblyAnnotation, "");
+	visit(*_assignment.value);
+	for (auto const& variable: _assignment.variableNames)
+		if (m_inlineAssemblyAnnotation->externalReferences.count(&variable))
+			if (auto const* declaration = dynamic_cast<VariableDeclaration const*>(m_inlineAssemblyAnnotation->externalReferences.at(&variable).declaration))
+				m_currentNode->variableOccurrences.emplace_back(
+					*declaration,
+					VariableOccurrence::Kind::Assignment,
+					variable.location
+				);
+}
+
+void ControlFlowBuilder::operator()(yul::FunctionCall const& _functionCall)
+{
+	using namespace yul;
+	solAssert(m_currentNode && m_inlineAssemblyAnnotation, "");
+	yul::ASTWalker::operator()(_functionCall);
+	if (_functionCall.functionName.name == "revert"_yulstring)
+	{
+		connect(m_currentNode, m_revertNode);
+		m_currentNode = newLabel();
+	}
+	else if (_functionCall.functionName.name == "return"_yulstring)
+	{
+		// TODO: how to handle return data? When to consider storage return variables as being assigned to and when not?
+		connect(m_currentNode, m_returnNode);
```

Comment:
If this is currently an internal solidity function, it does not even really return, it should be more treated like a 'revert'.

If it is a public or external function, we also cannot treat it as a solidity return, since it bypasses the modifiers and it can return just about anything - the return variables and the data area referenced in the return function call are two different things. I think we should treat it similar to 'revert', too, meaning: This bypasses the assignedness check, but creates errors for unreachable code. If this is an alternative branch and the other branch exits the function without assigning, it should still be an error. If all paths use this way to return and there is no assignment, it should not be an error.